### PR TITLE
macho: add TLS support

### PR DESCRIPTION
**Pull request by kubkon**

### CODE REVIEW

File: src/link/MachO.zig [hunk 1]
```diff
@@ -702,17 +702,23 @@ pub fn flushModule(self: *MachO, comp: *Compilation, prog_node: *std.Progress.No
...
 
-    if (self.unresolved.count() > 0) {
+    blk: {
+        for (self.unresolved.keys()) |index| {
+            // TODO: convert to compiler errors.
+            const global = self.globals.items[index];
+            const sym_name = self.getSymbolName(global);
+            log.err("undefined symbol reference '{s}'", .{sym_name});
+        } else break :blk;
         return error.UndefinedSymbolReference;
```
Comment (jacobly0) [inline src/link/MachO.zig:722]: There are no breaks out of the for loop, so the for else always runs and this return is unreachable.

Comment (kubkon) [reply src/link/MachO.zig:722]: Oh I see it now, yes, you are right. Thanks for pointing the one!

File: src/arch/x86_64/CodeGen.zig [hunk 2]
```diff
@@ -3772,13 +3799,17 @@ fn store(self: *Self, ptr: MCValue, value: MCValue, ptr_ty: Type, value_ty: Type
...
+            if (ptr.isThreadlocal()) {
+                try self.loadThreadLocalPtrIntoRegister(addr_reg, ptr_ty, ptr);
+            } else {
+                try self.loadMemPtrIntoRegister(addr_reg, ptr_ty, ptr);
+                // Load the pointer, which is stored in memory
+                try self.asmRegisterMemory(
+                    .mov,
+                    addr_reg.to64(),
+                    Memory.sib(.qword, .{ .base = addr_reg.to64() }),
+                );
```
Comment (jacobly0) [inline src/arch/x86_64/CodeGen.zig:3811]: Why would this not apply to thread local pointers?  `ptr` is a global pointer, so once you have a pointer to the global, you have to load the address out of memory.  Any instance where the `isThreadLocal` check can't be moved into `loadMemPtrIntoRegister` seems suspect.

File: src/arch/x86_64/CodeGen.zig [hunk 3]
```diff
@@ -7375,12 +7442,16 @@ fn genSetReg(self: *Self, ty: Type, reg: Register, mcv: MCValue) InnerError!void
...
-                    );
+                    if (mcv.isThreadlocal()) {
+                        try self.loadThreadLocalPtrIntoRegister(reg, Type.usize, mcv);
+                    } else {
+                        try self.loadMemPtrIntoRegister(reg, Type.usize, mcv);
+                        try self.asmRegisterMemory(
+                            .mov,
+                            registerAlias(reg, abi_size),
+                            Memory.sib(Memory.PtrSize.fromSize(abi_size), .{ .base = reg.to64() }),
+                        );
```
Comment (jacobly0) [inline src/arch/x86_64/CodeGen.zig:7453]: Same thing here, but this time we want to load a global value into a register, so once we have a pointer to the global, we need to load from it.

File: src/arch/x86_64/CodeGen.zig (hunk 2 above)
Comment (kubkon) [reply src/arch/x86_64/CodeGen.zig:3811]: We end up doing a double dereference at the wrong time. For standard GOT deref we do something like:

```
lea reg, [rip]
mov reg, [reg]
mov value_reg, [reg]
```

while for TLV we don't want the middle deref as we get the address of a thread-local variable:

```
mov rdi, [rip]
call [rdi]
mov reg, rax
mov value_reg, [reg]
```

**Comment by jacobly0** [comment]
I made my last review thinking it was based on a commit that I forgot wasn't upstream yet, so it probably doesn't make any sense.

**Comment by kubkon** [comment]
> I made my last review thinking it was based on a commit that I forgot wasn't upstream yet, so it probably doesn't make any sense.

No problem at all! I will rebase on the commit you are referring to and we can carry on from there. Thanks for your continuous feedback - it's been very useful!

**Pull request merged**